rasa/cdu/command_processor.py

```diff
@@ -144,20 +127,47 @@ def execute_commands(
 
     validate_state_of_commands(commands)
 
-    for i, command in enumerate(reversed_commands):
+    for command in reversed_commands:
         if isinstance(command, CorrectSlotsCommand):
+            if not current_top_flow:
+                # we shouldn't end up here as a correction shouldn't be triggered
+                # if there is nothing to correct. but just in case we do, we
+                # just skip the command.
+                structlogger.warning(
+                    "command_executor.correct_slots.no_active_flow", command=command
+                )
+                continue
             structlogger.debug("command_executor.correct_slots", command=command)
-            for correction in command.corrected_slots:
-                events.append(SlotSet(correction.name, correction.value))
-            events.append(
-                SlotSet(CORRECTED_SLOTS_SLOT, [s.name for s in command.corrected_slots])
+            proposed_slots = {c.name: c.value for c in command.corrected_slots}
+
+            reset_step = _find_earliest_updated_question(
+                user_step, user_flow, proposed_slots
             )
-            flow_stack.push(
-                FlowStackFrame(
-                    flow_id=FLOW_PATTERN_CORRECTION_ID,
-                    frame_type=StackFrameType.CORRECTION,
-                )
+            context = {
+                "corrected_slots": proposed_slots,
+                "corrected_reset_point": {
+                    "id": user_flow.id if user_flow else None,
+                    "step_id": reset_step.id if reset_step else None,
+                },
+            }
+            correction_frame = FlowStackFrame(
+                flow_id=FLOW_PATTERN_CORRECTION_ID,
+                frame_type=StackFrameType.REMARK,
+                context=context,
             )
+
+            if current_top_flow.id != FLOW_PATTERN_CORRECTION_ID:
+                flow_stack.push(correction_frame)
+            else:
+                # wrap up the previous correction flow
+                flow_stack.frames[-1].step_id = END_STEP
+                # push a new correction flow
+                flow_stack.push(
+                    correction_frame,
+                    # we allow the previous correction to finish first before
+                    # starting the new one
+                    index=-1,
+                )
```

Comment:
Actually, I think this would allow for attacking the flow diagram step logic. An "attack" could look like this:
1. go through the flow with values you think will satisfy conditions and get you to a desired leaf node
2. introduce some odd corrections (e.g. more money than you have, negative monetary value or sth)
3. introduce another correction before accepting the first one that is closest to your desired leaf node. 

Step two would introduce the bad data, and jump back. Before the steps are re-executed, step three would jump you back close to your desired leaf note, potentially skipping a lot of validations.

---
Even if it might be good to do a final validation before some API call, it seems we do want to rely on the flow logic validation quite a bit, as it is necessary for the branching logic. So we should probably address this. One potential solution would probably be merging the two frames ~ up to you if you want to do it here or in a follow up ticket i'd say

Reply:
maybe I also got it wrong as to what is currently happening here, yeah let's discuss